packages/aws-cdk/lib/api/hotswap/lambda-functions.ts

```diff
@@ -0,0 +1,159 @@
+      functionPhysicalName = undefined;
+    }
+
+    return new LambdaFunctionHotswapOperation({
+      logicalId,
+      physicalName: functionPhysicalName,
+      code: lambdaCodeChange,
+    });
+  }
+}
+
+/**
+ * Returns `true` if the change is not for a AWS::Lambda::Function,
+ * but doesn't prevent short-circuiting
+ * (like a change to CDKMetadata resource),
+ * `false` if the change is to a AWS::Lambda::Function,
+ * but not only to its Code property,
+ * or a LambdaFunctionCode if the change is to a AWS::Lambda::Function,
+ * and only affects its Code property.
+ */
+function isLambdaFunctionCodeOnlyChange(
+  change: cfn_diff.ResourceDifference, assetParamsWithEnv: { [key: string]: string },
+): LambdaFunctionCode | ChangeHotswapImpact {
+  if (!change.newValue) {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+  const newResourceType = change.newValue.Type;
+  // Ignore Metadata changes
+  if (newResourceType === 'AWS::CDK::Metadata') {
+    return ChangeHotswapImpact.IRRELEVANT;
+  }
+  // The only other resource change we should see is a Lambda function
+  if (newResourceType !== 'AWS::Lambda::Function') {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+  if (change.oldValue?.Type == null) {
+    // this means this is a brand-new Lambda function -
+    // obviously, we can't short-circuit that!
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  /*
+   * On first glance, we would want to initialize these using the "previous" values (change.oldValue),
+   * in case only one of them changed, like the key, and the Bucket stayed the same.
+   * However, that actually fails for old-style synthesis, which uses CFN Parameters!
+   * Because the names of the Parameters depend on the hash of the Asset,
+   * the Parameters used for the "old" values no longer exist in `assetParams` at this point,
+   * which means we don't have the correct values available to evaluate the CFN expression with.
+   * Fortunately, the diff will always include both the s3Bucket and s3Key parts of the Lambda's Code property,
+   * even if only one of them was actually changed,
+   * which means we don't need the "old" values at all, and we can safely initialize these with just `''`.
+   */
+  let s3Bucket = '', s3Key = '';
+  let foundCodeDifference = false;
+  // Make sure only the code in the Lambda function changed
+  const propertyUpdates = change.propertyUpdates;
+  for (const updatedPropName in propertyUpdates) {
+    const updatedProp = propertyUpdates[updatedPropName];
+    if (updatedProp.newValue === undefined) {
+      return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+    }
+    for (const newPropName in updatedProp.newValue) {
+      switch (newPropName) {
+        case 'S3Bucket':
+          foundCodeDifference = true;
+          s3Bucket = stringifyPotentialCfnExpression(updatedProp.newValue[newPropName], assetParamsWithEnv);
```

Comment:
Feels like you'd actually have to check whether this is actually the asset bucket or not. 

People could be referencing arbitrary crap but we can only hotswap if we actually end up uploading to the target bucket.

Reply:
In fact you could cross-reference this with a list of actually-uploaded assets, which we have because we just did the asset upload.

Reply:
> In fact you could cross-reference this with a list of actual assets, which we have because we just did the asset upload.

This is exactly what we do here with the `assetParamsWithEnv` parameter. If people are referencing arbitrary crap in their template, `stringifyPotentialCfnExpression()` will error out, and thus fail the entire `cdk deploy --hotswap` command, which is what we want here.